src/main/java/com/google/sps/servlets/StudyServlet.java

```diff
@@ -0,0 +1,162 @@
+          userCards.add(initializeCard(entity));
+        }
+      }
+      // Sorts cards in increasing order of familiarity score
+      UserCards.sort((Card c1, Card c2) -> Double.compare(c1.getFamiliarityScore(), c2.getFamiliarityScore()));
+      List<List<QuizCard>> quiz = createQuizRounds(userCards);
+
+      Gson gson = new Gson();
+      String jsonResponse = gson.toJson(quiz);
+      response.setContentType("application/json;");
+      response.getWriter().println(jsonResponse);
+    }
+  }
+
+  // Updates and stores a card's new familiarity score
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    Entity cardEntity;
+    Double newScore;
+    String answeredCorrectly = request.getParameter("answeredCorrectly");
+    String cardKey = request.getParameter("cardKey");
+    long time = System.currentTimeMillis();
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+
+    try {
+      cardEntity = datastore.get(KeyFactory.stringToKey(cardKey));
+    } catch (EntityNotFoundException e) {
+      cardEntity = null;
+    }
+
+    if (cardEntity != null) {
+      Card card = new Card(cardEntity);
+      if (answeredCorrectly.equals("true")) {
+        newScore = incFamiliarityScore(time, card.getFamiliarityScore(), card.getTimeTested());
+      } else {
+        newScore = decFamiliarityScore(time, card.getFamiliarityScore(), card.getTimeTested());
+      }
+      cardEntity.setProperty("familiarityScore", newScore);
+      cardEntity.setProperty("timeTested", time);
+      datastore.put(cardEntity);
+    }
+  }
+
+  /**
+   * Takes a list of cards that have been sorted and breaks them into smaller lists that represent
+   * rounds.
+   */
+  private List<List<QuizCard>> createQuizRounds(List<Card> userCards) {
+    List<List<QuizCard>> quiz = new ArrayList<List<QuizCard>>();
+    List<QuizCard> holder = new ArrayList<QuizCard>();
+    int numRounds = userCards.size() / numOfCardsPerRound;
+    int start = 0;
+    if (numRounds > maxNumOfRounds) {
+      numRounds = maxNumOfRounds;
+    }
+
+    for (int i = 0; i < numRounds; i++) {
+      quiz.add(new ArrayList<QuizCard>());
+      for (int j = start; j < (start + numOfCardsPerRound); j++) {
+        // Calls constructor that fetches similar words for the quiz
+        QuizCard quizCard = new QuizCard(userCards.get(j));
+        quiz.get(i).add(quizCard);
+      }
+      start += numOfCardsPerRound;
+    }
+
+    return quiz;
+  }
+
+  private Card initializeCard(Entity entity) {
+    // If cards don't have a familiarity score then provide default value
+    if (!entity.hasProperty("familiarityScore")) {
+      entity.setProperty("familiarityScore", .5);
+      entity.setProperty("timeTested", System.currentTimeMillis());
+      DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+      datastore.put(entity);
```

Comment:
I think it's confusing if side effects are in a helper function if that helper function's name doesn't clearly indicate that it performs side effects. So either change the name, or put this storing logic into the handler itself.

Also, you probably want to use a transaction here -- if the request fails, then you don't want this data to be persisted permaturely.

Reply:
done.